grid2op/Opponent/NeuripsOpponent.py

```diff
@@ -0,0 +1,146 @@
+            warnings.warn('The usage rate normalization is not specified. No normalization will be performed.')
+        elif len(rho_normalization) != len(lines_attacked):
+            raise Warning(f'The usage rate normalization must have the same length as the number '
+                          f'of attacked lines. No normalization will be performed.')
+        else:
+            self._rho_normalization = np.array(rho_normalization)
+
+        # Opponent's attack period
+        self._attack_period = attack_period
+
+    def reset(self, initial_budget):
+        self._next_attack_time = None
+
+    def tell_attack_continues(self, observation, agent_action, env_action, budget):
+        self._next_attack_time = None
+
+    def attack(self, observation, agent_action, env_action,
+               budget, previous_fails):
+        """
+        This method is the equivalent of "attack" for a regular agent.
+
+        Opponent, in this framework can have more information than a regular agent (in particular it can
+        view time step t+1), it has access to its current budget etc.
+
+        Parameters
+        ----------
+        observation: :class:`grid2op.Observation.Observation`
+            The last observation (at time t)
+
+        opp_reward: ``float``
+            THe opponent "reward" (equivalent to the agent reward, but for the opponent) TODO do i add it back ???
+
+        done: ``bool``
+            Whether the game ended or not TODO do i add it back ???
+
+        agent_action: :class:`grid2op.Action.Action`
+            The action that the agent took
+
+        env_action: :class:`grid2op.Action.Action`
+            The modification that the environment will take.
+
+        budget: ``float``
+            The current remaining budget (if an action is above this budget, it will be replaced by a do nothing.
+
+        previous_fails: ``bool``
+            Wheter the previous attack failed (due to budget or ambiguous action)
+
+        Returns
+        -------
+        attack: :class:`grid2op.Action.Action`
+            The attack performed by the opponent. In this case, a do nothing, all the time.
+        """
+        # TODO maybe have a class "GymOpponent" where the observation would include the budget  and all other
+        # TODO information, and forward something to the "act" method.
+
+        # During creation of the environment, do not attack
+        if observation is None:
+            return None
+
+        # Decide the time of the next attack
+        if self._next_attack_time is None:
+            self._next_attack_time = 1 + self.space_prng.randint(self._attack_period)
+        self._next_attack_time -= 1
+
+        # If the attack time has not come yet, do not attack
+        if self._next_attack_time > 0:
+            return None
+
+        # If all attackable lines are disconnected, do not attack
+        status = observation.line_status[self._lines_ids]
+        if np.all(status == False):
+            return None
```

Comment:
re sample the _next_attac_time in this case, no ?

Reply:
I don't think so. In that case where it is time to attack but all lines are disconnected, the opponent will simply attack as soon as a line is available to attack (connected).